### PR TITLE
improvement: retry on retryable bigquery exception: rateLimitExceeded…

### PR DESCRIPTION
**Pull request by tiboun**
Based on the following response from bigquery along with their exception:

```
{
  "code": 409,
  "errors": [
    {
      "domain": "global",
      "message": "Already Exists: Dataset ***",
      "reason": "duplicate"
    }
  ],
  "message": "Already Exists: Dataset ***",
  "status": "ALREADY_EXISTS"
}
```

```
{
  "code": 403,
  "errors": [
    {
      "domain": "usageLimits",
      "message": "Exceeded rate limits: too many dataset metadata update operations for this dataset. For more information, see https://cloud.google.com/bigquery/docs/troubleshoot-quotas",
      "reason": "rateLimitExceeded"
    }
  ],
  "message": "Exceeded rate limits: too many dataset metadata update operations for this dataset. For more information, see https://cloud.google.com/bigquery/docs/troubleshoot-quotas",
  "status": "PERMISSION_DENIED"
}
```

add a retry mechanism with backoff for rate limit since they are recoverable.

### CODE REVIEW

File: src/main/scala/ai/starlake/job/sink/bigquery/BigQueryJobBase.scala [hunk 1]
```diff
@@ -308,6 +308,25 @@ trait BigQueryJobBase extends StrictLogging {
 
   protected lazy val bqNativeTable: String = BigQueryJobBase.getBqNativeTable(tableId)
 
+  /** Retry on retryable bigquery exception.
+    */
+  private def recoverBigqueryException[T](bigqueryProcess: => T, retry: Int = 0): Try[T] = {
+    Try {
```
Comment (hayssams) [inline src/main/scala/ai/starlake/job/sink/bigquery/BigQueryJobBase.scala]: Did it work for you ?

**Review by hayssams — APPROVED** [review]

File: src/main/scala/ai/starlake/job/sink/bigquery/BigQueryJobBase.scala (hunk 1 above)
Comment (tiboun) [reply src/main/scala/ai/starlake/job/sink/bigquery/BigQueryJobBase.scala]: I've tested the duplicate case by forcing the call related to creation and it works but I didn't test the rate one yet. I plan to deploy a snapshot version and launch many ingestion process to see if failures happen again but it should not.

Comment (hayssams) [reply src/main/scala/ai/starlake/job/sink/bigquery/BigQueryJobBase.scala]: Ok let’s just merge this and est then 
Thank you

Comment (mhdkassir) [reply src/main/scala/ai/starlake/job/sink/bigquery/BigQueryJobBase.scala]: @hayssams @tiboun for info I'll push another PR next week to refactor some bigquery Api calls that seem to cause quota problems. We faced the issue and did a fix that is working

Comment (hayssams) [reply src/main/scala/ai/starlake/job/sink/bigquery/BigQueryJobBase.scala]: Ok great 👍

**Comment by tiboun** [comment]
While testing the fix above which works fine, I've encountered a swallowed exception. The output is a successful job and what is missing because of this swallowed exception is just the update of the columns and table descriptions in my case but we may miss other case if we swallowed it.

```
ERROR BigQuerySparkJob: com.google.cloud.bigquery.BigQueryException: Exceeded rate limits: too many table update operations for this table. For more information, see https://cloud.google.com/bigquery/docs/troubleshoot-quotas
	at com.google.cloud.bigquery.spi.v2.HttpBigQueryRpc.translate(HttpBigQueryRpc.java:114)
	at com.google.cloud.bigquery.spi.v2.HttpBigQueryRpc.patch(HttpBigQueryRpc.java:284)
	at com.google.cloud.bigquery.BigQueryImpl$15.call(BigQueryImpl.java:696)
	at com.google.cloud.bigquery.BigQueryImpl$15.call(BigQueryImpl.java:693)
	at com.google.api.gax.retrying.DirectRetryingExecutor.submit(DirectRetryingExecutor.java:103)
	at com.google.cloud.RetryHelper.run(RetryHelper.java:76)
	at com.google.cloud.RetryHelper.runWithRetries(RetryHelper.java:50)
	at com.google.cloud.bigquery.BigQueryImpl.update(BigQueryImpl.java:692)
	at ai.starlake.job.sink.bigquery.BigQueryJobBase.updateColumnsDescription(BigQueryJobBase.scala:547)
	at ai.starlake.job.sink.bigquery.BigQueryJobBase.updateColumnsDescription$(BigQueryJobBase.scala:528)
	at ai.starlake.job.sink.bigquery.BigQuerySparkJob.updateColumnsDescription(BigQuerySparkJob.scala:23)
	at ai.starlake.job.sink.bigquery.BigQuerySparkJob.$anonfun$runSparkConnector$7(BigQuerySparkJob.scala:274)
	at scala.util.Success.$anonfun$map$1(Try.scala:255)
	at scala.util.Success.map(Try.scala:213)
	at ai.starlake.job.sink.bigquery.BigQuerySparkJob.runSparkConnector(BigQuerySparkJob.scala:113)
	at ai.starlake.job.sink.bigquery.BigQuerySparkJob.run(BigQuerySparkJob.scala:287)
	at ai.starlake.job.ingest.IngestionUtil$.sinkRejected(IngestionJob.scala:1353)
	at ai.starlake.job.ingest.IngestionJob.saveRejected(IngestionJob.scala:153)
	at ai.starlake.job.ingest.IngestionJob.saveRejected$(IngestionJob.scala:124)
	at ai.starlake.job.ingest.JsonIngestionJob.saveRejected(JsonIngestionJob.scala:49)
	at ai.starlake.job.ingest.JsonIngestionJob.ingest(JsonIngestionJob.scala:162)
	at ai.starlake.job.ingest.IngestionJob.$anonfun$runSpark$2(IngestionJob.scala:1082)
	at scala.util.Try$.apply(Try.scala:213)
	at ai.starlake.job.ingest.IngestionJob.runSpark(IngestionJob.scala:1081)
	at ai.starlake.job.ingest.IngestionJob.runSpark$(IngestionJob.scala:1063)
	at ai.starlake.job.ingest.JsonIngestionJob.runSpark(JsonIngestionJob.scala:49)
	at ai.starlake.job.ingest.IngestionJob.run(IngestionJob.scala:1014)
	at ai.starlake.job.ingest.IngestionJob.run$(IngestionJob.scala:1009)
	at ai.starlake.job.ingest.JsonIngestionJob.run(JsonIngestionJob.scala:49)
	at ai.starlake.workflow.IngestionWorkflow.$anonfun$ingest$3(IngestionWorkflow.scala:516)
	at scala.util.Try$.apply(Try.scala:213)
	at ai.starlake.workflow.IngestionWorkflow.ingest(IngestionWorkflow.scala:472)
	at ai.starlake.schema.handlers.SimpleLauncher.ingest(LaunchHandler.scala:134)
	at ai.starlake.workflow.IngestionWorkflow.$anonfun$loadPending$17(IngestionWorkflow.scala:324)
	at ai.starlake.workflow.IngestionWorkflow.$anonfun$loadPending$17$adapted(IngestionWorkflow.scala:323)
	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:286)
	at scala.collection.immutable.List.foreach(List.scala:431)
	at scala.collection.TraversableLike.map(TraversableLike.scala:286)
	at scala.collection.TraversableLike.map$(TraversableLike.scala:279)
	at scala.collection.immutable.List.map(List.scala:305)
	at ai.starlake.workflow.IngestionWorkflow.$anonfun$loadPending$14(IngestionWorkflow.scala:323)
	at ai.starlake.workflow.IngestionWorkflow.$anonfun$loadPending$14$adapted(IngestionWorkflow.scala:306)
	at scala.collection.Iterator$$anon$10.next(Iterator.scala:461)
	at scala.collection.Iterator.foreach(Iterator.scala:943)
	at scala.collection.Iterator.foreach$(Iterator.scala:943)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1431)
	at scala.collection.generic.Growable.$plus$plus$eq(Growable.scala:62)
	at scala.collection.generic.Growable.$plus$plus$eq$(Growable.scala:53)
	at scala.collection.mutable.ListBuffer.$plus$plus$eq(ListBuffer.scala:184)
	at scala.collection.mutable.ListBuffer.$plus$plus$eq(ListBuffer.scala:47)
	at scala.collection.generic.GenericTraversableTemplate.$anonfun$flatten$1(GenericTraversableTemplate.scala:174)
	at scala.collection.immutable.List.foreach(List.scala:431)
	at scala.collection.generic.GenericTraversableTemplate.flatten(GenericTraversableTemplate.scala:173)
	at scala.collection.generic.GenericTraversableTemplate.flatten$(GenericTraversableTemplate.scala:171)
	at scala.collection.AbstractTraversable.flatten(Traversable.scala:108)
	at ai.starlake.workflow.IngestionWorkflow.loadPending(IngestionWorkflow.scala:341)
	at ai.starlake.job.Main.run(Main.scala:209)
	at ai.starlake.job.Main$.main(Main.scala:59)
	at ai.starlake.job.Main.main(Main.scala)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.apache.spark.deploy.JavaMainApplication.start(SparkApplication.scala:52)
	at org.apache.spark.deploy.SparkSubmit.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:958)
	at org.apache.spark.deploy.SparkSubmit.doRunMain$1(SparkSubmit.scala:180)
	at org.apache.spark.deploy.SparkSubmit.submit(SparkSubmit.scala:203)
	at org.apache.spark.deploy.SparkSubmit.doSubmit(SparkSubmit.scala:90)
	at org.apache.spark.deploy.SparkSubmit$$anon$2.doSubmit(SparkSubmit.scala:1046)
	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:1055)
	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
Caused by: com.google.api.client.googleapis.json.GoogleJsonResponseException: 403 Forbidden
POST https://www.googleapis.com/bigquery/v2/projects/***/datasets/audit/tables/rejected?prettyPrint=false
{
  "code": 403,
  "errors": [
    {
      "domain": "usageLimits",
      "message": "Exceeded rate limits: too many table update operations for this table. For more information, see https://cloud.google.com/bigquery/docs/troubleshoot-quotas",
      "reason": "rateLimitExceeded"
    }
  ],
  "message": "Exceeded rate limits: too many table update operations for this table. For more information, see https://cloud.google.com/bigquery/docs/troubleshoot-quotas",
  "status": "PERMISSION_DENIED"
}
	at com.google.api.client.googleapis.json.GoogleJsonResponseException.from(GoogleJsonResponseException.java:146)
	at com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest.newExceptionOnError(AbstractGoogleJsonClientRequest.java:118)
	at com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest.newExceptionOnError(AbstractGoogleJsonClientRequest.java:37)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest$3.interceptResponse(AbstractGoogleClientRequest.java:466)
	at com.google.api.client.http.HttpRequest.execute(HttpRequest.java:1111)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:552)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:493)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.execute(AbstractGoogleClientRequest.java:603)
	at com.google.cloud.bigquery.spi.v2.HttpBigQueryRpc.patch(HttpBigQueryRpc.java:282)
	... 69 more

23/06/13 13:31:42 ERROR JsonIngestionJob: Failed to save Rejected
```

I'll throw it again if you don't mind.

@mhdkassir have you changed the updateColumnDescription in order to limit the number of call to the API that may throw a quotas problem ? If not, I may implement something that will compare the column descriptions before applying any changes like what has been done in `getOrCreateDataset`. What do you think ?

**Comment by tiboun** [comment]
Maybe I'm wrong but It seems that BigQueryJobBase#updateColumnsDescription update root's column description only, isn't it ?

**Comment by hayssams** [comment]
> Maybe I'm wrong but It seems that BigQueryJobBase#updateColumnsDescription update root's column description only, isn't it ?

@mhdkassir I think that this is question for you.
@tiboun Yes I think so.

**Comment by tiboun** [comment]
@hayssams I've tested succesfully this change but didn't have a use case to get rate limite exception yet. Can you please check again this PR with the additionnal improvements ?

File: src/main/scala/ai/starlake/job/ingest/DsvIngestionJob.scala [hunk 2]
```diff
@@ -205,8 +205,13 @@ class DsvIngestionJob(
       mergedMetadata.emptyIsNull.getOrElse(settings.comet.emptyIsNull)
     )
 
-    saveRejected(validationResult.errors, validationResult.rejected)
-    saveAccepted(validationResult)
+    saveRejected(validationResult.errors, validationResult.rejected).map { _ =>
```
Comment (hayssams) [inline src/main/scala/ai/starlake/job/ingest/DsvIngestionJob.scala:208]: OK Gr8 to raise an exception here

File: src/main/scala/ai/starlake/job/ingest/GenericIngestionJob.scala [hunk 3]
```diff
@@ -308,8 +308,13 @@ class GenericIngestionJob(
       mergedMetadata.emptyIsNull.getOrElse(settings.comet.emptyIsNull)
     )
 
-    saveRejected(validationResult.errors, validationResult.rejected)
-    saveAccepted(validationResult)
+    saveRejected(validationResult.errors, validationResult.rejected).map { _ =>
```
Comment (hayssams) [inline src/main/scala/ai/starlake/job/ingest/GenericIngestionJob.scala:311]: ok

File: src/main/scala/ai/starlake/job/ingest/JsonIngestionJob.scala [hunk 4]
```diff
@@ -159,8 +159,13 @@ class JsonIngestionJob(
     import session.implicits._
     val rejectedDS = session.createDataset(withInvalidSchema).union(validationResult.errors)
 
-    saveRejected(rejectedDS, validationResult.rejected)
-    saveAccepted(validationResult) // prefer to let Spark compute the final schema
+    saveRejected(rejectedDS, validationResult.rejected).map { _ =>
```
Comment (hayssams) [inline src/main/scala/ai/starlake/job/ingest/JsonIngestionJob.scala:162]: ok

File: src/main/scala/ai/starlake/job/ingest/ParquetIngestionJob.scala [hunk 5]
```diff
@@ -137,8 +137,13 @@ class ParquetIngestionJob(
       mergedMetadata.emptyIsNull.getOrElse(settings.comet.emptyIsNull)
     )
 
-    saveRejected(validationResult.errors, validationResult.rejected)
-    saveAccepted(validationResult)
+    saveRejected(validationResult.errors, validationResult.rejected).map { _ =>
```
Comment (hayssams) [inline src/main/scala/ai/starlake/job/ingest/ParquetIngestionJob.scala:140]: ok

File: src/main/scala/ai/starlake/job/ingest/PositionIngestionJob.scala [hunk 6]
```diff
@@ -114,8 +114,13 @@ class PositionIngestionJob(
       settings.comet.sinkReplayToFile,
       mergedMetadata.emptyIsNull.getOrElse(settings.comet.emptyIsNull)
     )
-    saveRejected(validationResult.errors, validationResult.rejected)
-    saveAccepted(validationResult)
+    saveRejected(validationResult.errors, validationResult.rejected).map { _ =>
```
Comment (hayssams) [inline src/main/scala/ai/starlake/job/ingest/PositionIngestionJob.scala:117]: ok

**Comment by hayssams** [comment]
@tiboun Just checked your PR.
It is definitely a huge improvement over the previous version.

**Comment by tiboun** [comment]
Thanks @hayssams !

**Comment by mhdkassir** [comment]
> > Maybe I'm wrong but It seems that BigQueryJobBase#updateColumnsDescription update root's column description only, isn't it ?
> 
> @mhdkassir I think that this is question for you. @tiboun Yes I think so.

@tiboun  yes I confirm as well.

**Pull request merged**